pkg/status/cel-evaluator.go

```diff
@@ -0,0 +1,111 @@
+/*
+Copyright 2024 The KubeStellar Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package status
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/google/cel-go/cel"
+	"github.com/google/cel-go/checker/decls"
+	"github.com/google/cel-go/common/types/ref"
+
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+// celEvaluator is a struct that holds the CEL environment
+// and provides a method to evaluate an expression with an unstructured object
+// as the context.
+type celEvaluator struct {
+	env *cel.Env
+}
+
+// NewCELEvaluator initializes the CEL environment.
+func newCELEvaluator() (*celEvaluator, error) {
+	env, err := cel.NewEnv(
+		cel.Declarations(
+			decls.NewVar("obj", decls.NewMapType(decls.String, decls.Dyn)),
+		),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create CEL environment: %v", err)
+	}
+
+	return &celEvaluator{env: env}, nil
+}
+
+// preprocessExpression replaces occurrences of $ with obj.
+func (e *celEvaluator) preprocessExpression(expression string) string {
+	return strings.ReplaceAll(expression, "$", "obj")
```

Comment:
Hmm, did we never write down concrete `Expression` values?

Certainly a string-level replacement of dollar-sign with "obj" can not be valid, we should allow expressions like `something == "$"`, for example.

I was supposing that that accessing a field like `spec.replicas` would be done with an `Expression` like `spec.replicas`, for simplicity and conciseness. We _could_ instead define one CEL "variable" that holds the whole object. It is worth writing the decision down clearly somewhere prominent.